### PR TITLE
Trim whitespace in create how-to

### LINKED ISSUE #2635: [bug] whitespace not stripped on howto form inputs
**Describe the bug**
How-to `title`, `step description` and `step title` allow whitespace inputs. Leading and trailing whitespace count toward the character count, and validation passes when the user enters a bunch of spaces.

**To Reproduce**
Steps to reproduce the behaviour:

1. Go to `http://localhost:3000/how-to/create`
2. Fill out the form entirely so that it is valid (except the how-to title, step title, and step description)
3. Enter whitespace in the how-to title, step title, and step description until the form is valid
4. Save as draft, or publish. The save will succeed, but it should not.

**Expected behaviour**
We should validate against a trimmed input, and save a trimmed input. However I don't think we should require the user to manually delete any leading/trailing spaces (i.e. a valid input is still valid even if there are trailing spaces, as we will strip them on save)

**Screenshots**
Blank how-to title:
image

Here is a step that is filled entirely with spaces:
image

**Additional context**
- Does the server provide redundant validation?

### PR DESCRIPTION
**Pull request by CrowsVeldt**
PR Checklist

- [ ] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

_What this PR does_

Trim whitespace in how-to title, step title, and step description by adding a 'trim' modifier to FieldInput and FieldTextarea.

I messed up committing the last changes, so there are redundant commits. I'm still fairly new with git and I didn't want to mess it up more than I already had. If it's a problem I am happy to correct it, though I hope someone could give me some pointers.

## Git Issues

Closes #2635

## Screenshots/Videos

_If useful, provide screenshot or capture to highlight main changes_

---

## What happens next?

Thanks for the contribution! We try to make sure all PRs are reviewed ahead of a monthly dev call (first Monday of the month, open to all!).

If the PR is working as intended it'll be merged and included in the next platform release, if not changes will be requested and re-reviewed once updated.

If you need more immediate feedback you can try reaching out on Discord in the [Community Platform `development` channel](https://discord.com/channels/586676777334865928/938781727017558018).

### CODE REVIEW

**Comment by asdFletcher** [comment]
You can consolidate your commits into one using `rebase`, which is a great git tool to learn. If you're working within your own repo there shouldn't be any chance of any destructive action on the main repo. 

First, I'd checkout a backup branch with all your work in it. Then switch to the old branch, and run:

`git rebase -i main`

A new VSCode window should open with:

```sh
pick 13d06f8 commit 1 message
pick 6feb06e commit 2 message
pick 6feb06f commit 3 message
```
You can change it to:

```sh
pick 13d06f8 commit 1 message
f 6feb06e commit 2 message
f 6feb06f commit 3 message
```

And then close the window. Which will "squash" the 2nd and 3rd commits into the first.

You can do a diff with your backup branch, i think, which should be blank to verify nothing has changed. 

I'm on discord if you get stuck, and [the docs](https://git-scm.com/docs/git-rebase) are useful

**Comment by CrowsVeldt** [comment]
Fingers crossed I *think* I've cleaned up the commit history

**Comment by AlfonsoGhislieri** [comment]
Thanks for raising this PR @CrowsVeldt, would be great if you could add some tests to make sure this works as intended and to prevent any issues from arising in the future!

**Comment by CrowsVeldt** [comment]
This is my first time with cypress. Let me know if there are any changes or additions requires.

**Comment by CrowsVeldt** [comment]
I notice that FieldInput has a `type` for InputModifiers:
`type InputModifiers = {
  capitalize?: boolean
}`

where FieldTextarea does not. I was thinking of just copying it over, does that sound reasonably inside the scope of this PR?

Alternatively should it be moved to a separate file from which both FieldInput and FieldTextarea could inherit?

**Comment by AlfonsoGhislieri** [comment]
Hi @CrowsVeldt, sorry I never followed up on this, I went on holiday for a little month.. I'll get this all checked out next week when I'm back home! :)

**Comment by CrowsVeldt** [comment]
> Hi @CrowsVeldt, sorry I never followed up on this, I went on holiday for a little month.. I'll get this all checked out next week when I'm back home! :)

Thanks!

File: packages/cypress/src/integration/howto/write.spec.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import {
   HOWTO_TITLE_MIN_LENGTH,
 } from '../../../../../src/pages/Howto/constants'
 
+// ZECHY: Learn this shit
 describe('[How To]', () => {
```
Comment (AlfonsoGhislieri) [inline packages/cypress/src/integration/howto/write.spec.ts]: Could we remove this comment please :)

File: packages/cypress/src/integration/howto/write.spec.ts [hunk 2]
```diff
@@ -78,6 +77,22 @@ describe('[How To]', () => {
         .blur({ force: true })
 
       cy.get('[data-cy=step-description]').should('have.value', description)
+
+      // add test for trimming whitespace from step description
+
```
Comment (AlfonsoGhislieri) [inline packages/cypress/src/integration/howto/write.spec.ts]: Similarly let's remove this please!

**Comment by CrowsVeldt** [comment]
> * Cases with extra spaces in the message that are not leading or trailing eg: 'single   spaces'
> * Cases where there are no leading, trailing or extra spaces at all eg: 'normal message'

I'll see what I can do ;)

**Comment by CrowsVeldt** [comment]
I added tests for capitalizing the first letter, which I think also covers the normal case. Should I add a 'normal message' test specifically for checking that there's no extra whitespace? I feel like that would add more clutter than clarity.

As for extra whitespace inside a message, there's nothing currently in place to remove any. Should I add a test to checks if extra whitespace is removed, that fails? I am willing to try and implement removing extra whitespace inside the message, but it might take some time.

> Thanks for all of this @CrowsVeldt, I would say let's just maybe add two more tests to get a full coverage for all cases:
> 
> * Cases with extra spaces in the message that are not leading or trailing eg: 'single   spaces'
> * Cases where there are no leading, trailing or extra spaces at all eg: 'normal message'
> 
> This will give us robust coverage moving forwards :)
> 
> Edit: also would be good to solve the conflicts please!

**Comment by AlfonsoGhislieri** [comment]
Firstly, thanks very much for all the hard work and apologies for not being super responsive!

> Should I add a test to checks if extra whitespace is removed, that fails? I am willing to try and implement removing extra whitespace inside the message, but it might take some time.
> 
> >

Ah sorry I think this was me just being confusing, I was suggesting adding a test to precisely check that the extra whitespace is NOT removed when it's not trailing or leading if that makes sense? Just to make sure what you added isn't changing anything it shouldn't be! No need to add anything else for the scope of this PR. 

Also I saw we have some linting failures on the tests, please make sure you have prettier (https://prettier.io/) and make sure it's running the auto formatting :-)

Nearly there and thanks again, great work!

File: packages/cypress/src/integration/howto/write.spec.ts [hunk 3]
```diff
@@ -31,16 +31,69 @@ describe('[How To]', () => {
...
         .clear()
-        .invoke('val', description)
+        .type(description)
         .blur({ force: true })
 
+      cy.get('[data-cy=step-description]').should('have.value', description)
+
+      cy.get('[data-cy=character-count]')
+        .should('be.visible')
+        .contains(`101 / ${HOWTO_STEP_DESCRIPTION_MAX_LENGTH}`)
```
Comment (AlfonsoGhislieri) [inline packages/cypress/src/integration/howto/write.spec.ts:75]: **Issue** Just noticed that this was added, this test is causing the failures as the description lengths that are passed for these step tests are different lenghts (Not always 101, in some tests I believe it goes to the max 1000). Also while a good idea this is outside the scope of this issue!

File: packages/cypress/src/integration/howto/write.spec.ts [hunk 4]
```diff
@@ -31,16 +31,69 @@ describe('[How To]', () => {
...
+      cy.get('[data-cy=step-description]')
         .clear()
         .invoke('val', title)
         .blur({ force: true })
 
       cy.get('[data-cy=step-description]')
         .clear()
-        .invoke('val', description)
+        .type(description)
         .blur({ force: true })
```
Comment (AlfonsoGhislieri) [inline packages/cypress/src/integration/howto/write.spec.ts:69]: **Issue**: can we keep the original invoke function, since some descriptions are 1000 chars long this makes the tests take so much longer

Comment (CrowsVeldt) [reply packages/cypress/src/integration/howto/write.spec.ts:69]: Good to know. I didn't think of that

File: packages/cypress/src/integration/howto/write.spec.ts (hunk 3 above)
Comment (CrowsVeldt) [reply packages/cypress/src/integration/howto/write.spec.ts:75]: I'm pretty sure I didn't add this. I'm not sure how it ended up being marked as my new code. Maybe I messed up merging with master at some point?
Should I remove it?

Comment (AlfonsoGhislieri) [reply packages/cypress/src/integration/howto/write.spec.ts:75]: Yes please, yes probably came in from a merge!

**Comment by AlfonsoGhislieri** [comment]
Also some tests are now failing due to the commit messages, please make sure they adhere to [conventional commits ](https://www.conventionalcommits.org/en/v1.0.0/) 

In order to go back and fix these you can use `git rebase`: https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/changing-a-commit-message#

Then just make sure to force push :)

**Comment by AlfonsoGhislieri** [comment]
Little bit of a commit bomb here haha @CrowsVeldt, if it's alright with you I'll jump in here and clean this up a little/fix the failing tests over the weekend when I have some time? I'll raise another PR building off of yours, your commits and contribution will still be tracked, so you will still have full credit for the hard work! :-)

**Comment by CrowsVeldt** [comment]
> Little bit of a commit bomb here haha @CrowsVeldt, if it's alright with you I'll jump in here and clean this up a little/fix the failing tests over the weekend when I have some time? I'll raise another PR building off of yours, your commits and contribution will still be tracked, so you will still have full credit for the hard work! :-)

Please do, thank you! I'm still new at commit history management, I appreciate the support and encouragement :)

**Comment by AlfonsoGhislieri** [comment]
No worries, can definitely be a little daunting and confusing at the start. If you ever have any questions feel free to ask!

**Comment by CrowsVeldt** [comment]
> Hey @CrowsVeldt, I tried to poke around and it was just quite a mess as the code is quite stale compared to our most current production code (my fault for having be away and not helped with this back when you originally raised this). Could you raise a new PR where you implement these changes from our latest up to date code, hopefully shouldn't take too long? Thanks, really appreciate it :-)

done
